drivers/gpio/gpio_npm6001.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_npm6001_gpio
+
+#include <errno.h>
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/dt-bindings/gpio/nordic-npm6001-gpio.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/toolchain.h>
+
+/* nPM6001 GPIO related registers */
+#define NPM6001_GPIOOUTSET 0x69U
+#define NPM6001_GPIOOUTCLR 0x6AU
+#define NPM6001_GPIOIN	   0x6BU
+#define NPM6001_GPIO0CONF  0x6CU
+#define NPM6001_GPIO1CONF  0x6DU
+#define NPM6001_GPIO2CONF  0x6EU
+
+/* GPIO(0|1|2)CONF fields */
+#define NPM6001_CONF_DIRECTION_OUTPUT BIT(0)
+#define NPM6001_CONF_INPUT_ENABLED    BIT(1)
+#define NPM6001_CONF_PULLDOWN_ENABLED BIT(2)
+#define NPM6001_CONF_DRIVE_HIGH	      BIT(5)
+#define NPM6001_CONF_SENSE_CMOS	      BIT(6)
+
+#define NPM6001_PIN_MAX 2U
+#define NPM6001_PIN_MSK 0x7U
+
+struct gpio_npm6001_config {
+	struct gpio_driver_config common;
+	struct i2c_dt_spec bus;
+};
+
+struct gpio_npm6001_data {
+	struct gpio_driver_data common;
+};
+
+static int gpio_npm6001_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_npm6001_config *config = dev->config;
+	uint8_t reg = NPM6001_GPIOIN;
+	uint8_t val;
+	int ret;
+
```

Comment:
Perhaps guard these against ISR usage since the regular GPIO controllers can be used from ISRs? I have done that for the LMP90xxx GPIO controller, which sits on SPI. Something along the lines of:

```
    if (k_is_in_isr()) {
        return -EWOULDBLOCK;
    }
```

Reply:
will add, thanks!